sarif/issues_report.py

```diff
@@ -0,0 +1,146 @@
+                key_and_count = code_to_key_and_count.get(code, None)
+                if key_and_count is None:
+                    code_to_key_and_count[code] = {
+                        "key": key,
+                        "common_desc": record["Description"],
+                        "count": 1,
+                    }
+                else:
+                    key_and_count["count"] += 1
+                    common_desc_stem = key_and_count["common_desc"]
+                    desc = record["Description"]
+                    if common_desc_stem != desc:
+                        for char_pos, (char1, char2) in enumerate(
+                            zip(common_desc_stem, desc)
+                        ):
+                            if char1 != char2:
+                                key_and_count["key"] = (
+                                    common_desc_stem[0:char_pos] + " …"
+                                )
+                                break
+            sorted_codes = sorted(
+                code_to_key_and_count.keys(),
+                key=lambda code: -code_to_key_and_count[code]["count"],
+            )
+            self._sev_to_sorted_keys[severity] = {
+                code_to_key_and_count[code]["key"]: [] for code in sorted_codes
+            }
+            for record in issues:
+                # Not sorting the issues by location at this point
+                code = record["Code"]
+                self._sev_to_sorted_keys[severity][
+                    code_to_key_and_count[code]["key"]
+                ].append(record)
+
+    def _sort_record_lists(self):
+        if self._sev_to_sorted_keys is None:
+            self._group_records_by_key()
+        for key_to_records in self._sev_to_sorted_keys.values():
+            for records in key_to_records.values():
+                records.sort(key=record_sort_key)
+        self._records_have_been_sorted = True
+
+    def issue_count_for_severity(self, severity: str) -> int:
+        """Get the number of individual records at this severity level."""
+        return len(self._sev_to_records.get(severity, []))
+
+    def issue_type_count_for_severity(self, severity: str) -> int:
+        """Get the number of distinct issue types at this severity level."""
+        if self._sev_to_sorted_keys is None:
+            self._group_records_by_key()
+        return len(self._sev_to_sorted_keys.get(severity, []))
+
+    def any_none_severities(self) -> bool:
+        """Are there any records with severity level "none"?"""
+        return bool(self._sev_to_records.get("none", {}))
+
+    def get_severities(self) -> List[str]:
+        """
+        Get the list of relevant severity levels for these records.
+
+        The returned list always includes "error", "warning" and "note", but "none" is only
+        included at the end if there are any records with severity "none".
+        """
+        return (
+            SARIF_SEVERITIES_WITH_NONE
```

Comment:
Historically, the tools have always included error, warning and none regardless of counts.  There's an aspect of user education here as it immediately shows what the severity levels are, and also the satisfaction and clarity of seeing "errors: 0" rather than the section just disappearing.

But now we're adding "none" severity, we've deviated from the fixed list a bit, because it's a weird severity that most tools won't produce, so we don't want to include it by default.  I think the arguments for always including the other three still stand though.

I'll add some comments to explain